File: articles/disability-neurodivergence-and-development/python/model_disability_neurodivergence.py
#!/usr/bin/env python3
"""Model synthetic disability, neurodivergence, access, and participation."""

from __future__ import annotations

from pathlib import Path
import matplotlib.pyplot as plt
import pandas as pd
import statsmodels.formula.api as smf

ROOT = Path(__file__).resolve().parents[1]
DATA_PATH = ROOT / "data" / "disability_neurodivergence_panel.csv"
OUTPUTS_DIR = ROOT / "outputs"
OUTPUTS_DIR.mkdir(parents=True, exist_ok=True)


def main() -> None:
    if not DATA_PATH.exists():
        raise FileNotFoundError("Run python/generate_disability_neurodivergence_panel.py first.")

    panel = pd.read_csv(DATA_PATH).sort_values(["child_id", "time"]).reset_index(drop=True)
    panel["lag_score"] = panel.groupby("child_id")["development_score"].shift(1)
    model_data = panel.dropna(subset=["lag_score"]).copy()

    development_model = smf.ols(
        """
        development_score ~ lag_score + time + neuro_profile +
        current_support + current_access + current_barrier +
        current_communication + participation_score + current_advocacy +
        inclusion_climate + service_access + sensory_flexibility +
        current_support:current_access + current_barrier:neuro_profile
        """,
        data=model_data,
    ).fit(cov_type="HC3")

    participation_model = smf.ols(
        """
        participation_score ~ time + neuro_profile +
        current_support + current_access + current_barrier +
        current_communication + current_advocacy +
        inclusion_climate + service_access + sensory_flexibility
        """,
        data=panel,
    ).fit(cov_type="HC3")

    with open(OUTPUTS_DIR / "python_accessibility_model_summary.txt", "w", encoding="utf-8") as f:
        f.write("DYNAMIC DEVELOPMENT MODEL\n")
        f.write("=" * 80 + "\n")
        f.write(development_model.summary().as_text())
        f.write("\n\nPARTICIPATION MODEL\n")
        f.write("=" * 80 + "\n")
        f.write(participation_model.summary().as_text())

    trajectory = panel.groupby("time", as_index=False).agg(
        average_development=("development_score", "mean"),
        average_participation=("participation_score", "mean"),
        standard_error=("development_score", lambda x: x.std() / (len(x) ** 0.5)),
    )
    trajectory["lower"] = trajectory["average_development"] - 1.96 * trajectory["standard_error"]
    trajectory["upper"] = trajectory["average_development"] + 1.96 * trajectory["standard_error"]
    trajectory.to_csv(OUTPUTS_DIR / "python_development_trajectory.csv", index=False)

    plt.figure(figsize=(8, 5))
    plt.plot(trajectory["time"], trajectory["average_development"], linewidth=2)
    plt.fill_between(trajectory["time"], trajectory["lower"], trajectory["upper"], alpha=0.2)
    plt.xlabel("Time")
    plt.ylabel("Average development score")
    plt.title("Synthetic Disability, Neurodivergence, and Development")
    plt.tight_layout()
    plt.savefig(OUTPUTS_DIR / "python_development_trajectory.png", dpi=160)
    plt.close()

    plt.figure(figsize=(8, 5))
    plt.plot(trajectory["time"], trajectory["average_participation"], marker="o")
    plt.xlabel("Time")
    plt.ylabel("Average participation score")
    plt.title("Synthetic Participation Trajectory")
    plt.tight_layout()
    plt.savefig(OUTPUTS_DIR / "python_participation_trajectory.png", dpi=160)
    plt.close()

    profile_summary = panel.groupby("access_condition", as_index=False).agg(
        children=("child_id", "nunique"),
        average_development=("development_score", "mean"),
        average_participation=("participation_score", "mean"),
        average_access=("current_access", "mean"),
        average_barrier=("current_barrier", "mean"),
        average_support=("current_support", "mean"),
    )
    profile_summary.to_csv(OUTPUTS_DIR / "python_access_profiles.csv", index=False)

    print("Wrote Python accessibility model outputs.")


if __name__ == "__main__":
    main()

File: articles/disability-neurodivergence-and-development/r/model_disability_neurodivergence.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(dplyr)
  library(ggplot2)
  library(lme4)
})

args <- commandArgs(trailingOnly = FALSE)
file_arg <- "--file="
script_path <- normalizePath(sub(file_arg, "", args[grep(file_arg, args)]))
root <- normalizePath(file.path(dirname(script_path), ".."), mustWork = TRUE)

data_path <- file.path(root, "data", "disability_neurodivergence_panel.csv")
outputs_dir <- file.path(root, "outputs")

if (!dir.exists(outputs_dir)) dir.create(outputs_dir, recursive = TRUE)
if (!file.exists(data_path)) stop("Run python/generate_disability_neurodivergence_panel.py first.")

panel <- read.csv(data_path)

development_model <- lmer(
  development_score ~ time + neuro_profile + current_support +
    current_access + current_communication + current_barrier +
    participation_score + current_advocacy + inclusion_climate +
    service_access + sensory_flexibility +
    current_support:current_access +
    current_barrier:neuro_profile +
    (1 + time | setting_id/child_id),
  data = panel
)

participation_model <- lmer(
  participation_score ~ time + neuro_profile + current_support +
    current_access + current_communication + current_barrier +
    current_advocacy + inclusion_climate + service_access +
    sensory_flexibility +
    (1 + time | setting_id/child_id),
  data = panel
)

capture.output(
  list(
    development_model = summary(development_model),
    participation_model = summary(participation_model)
  ),
  file = file.path(outputs_dir, "r_accessibility_model_summary.txt")
)

trajectory <- panel |>
  group_by(time) |>
  summarize(
    average_development = mean(development_score),
    average_participation = mean(participation_score),
    standard_error = sd(development_score) / sqrt(n()),
    lower = average_development - 1.96 * standard_error,
    upper = average_development + 1.96 * standard_error,
    .groups = "drop"
  )

write.csv(trajectory, file.path(outputs_dir, "r_development_trajectory.csv"), row.names = FALSE)

ggsave(
  filename = file.path(outputs_dir, "r_development_trajectory.png"),
  plot = ggplot(trajectory, aes(x = time, y = average_development)) +
    geom_line(linewidth = 1) +
    geom_ribbon(aes(ymin = lower, ymax = upper), alpha = 0.15) +
    labs(title = "Synthetic Disability, Neurodivergence, and Development", x = "Time", y = "Average development") +
    theme_minimal(),
  width = 8,
  height = 5,
  dpi = 160
)

ggsave(
  filename = file.path(outputs_dir, "r_participation_trajectory.png"),
  plot = ggplot(trajectory, aes(x = time, y = average_participation)) +
    geom_line(linewidth = 1) +
    geom_point() +
    labs(title = "Synthetic Participation Trajectory", x = "Time", y = "Average participation") +
    theme_minimal(),
  width = 8,
  height = 5,
  dpi = 160
)

message("Wrote R accessibility outputs.")
